File: leetcode/backtracking/332. Reconstruct Itinerary/py/main.py
import collections

# Time: O(E^D) where |E| is the number of total flights and |D| is the maximum number of flights from an airport.
# Space: O(|V| + |E|) where |V| is the number of airports and |E| is the number of flights.
class Solution:
    def findItinerary(self, tickets):
        adList = collections.defaultdict(list)

        for f, t in tickets:
          adList[f].append(t)

        ans, visitBitmap = [], {}
        
        for k in adList:
          adList[k].sort()
          visitBitmap[k] = [False] * len(adList[k])

        def dfs(u, comb): 
          nonlocal ans
          
          if len(comb) == len(tickets) + 1:
            ans = list(comb)
            return True

          for i, v in enumerate(adList[u]):
            if not visitBitmap[u][i]:
                
                visitBitmap[u][i] = True
                comb.append(v)
                
                ret = dfs(v, comb)
                  
                visitBitmap[u][i] = False
                comb.pop()
                
                if ret:
                    return True
                
          return False
            
        dfs('JFK', ['JFK'])
        
        return ans

File: leetcode/backtracking/332. Reconstruct Itinerary/js/main.js
// Time: O(E^D)
// Space: O(∣V∣+∣E∣) where |V| is the number of airports and |E| is the number of flights.
const findItinerary = function (tickets) {
  if (tickets === null || tickets.length === 0) {
    return [];
  }

  let map = {};

  tickets.forEach(ticket => {
    const [from, to] = ticket;
    if (!map[from]) {
      map[from] = [to];
    } else {
      map[from].push(to);
      map[from].sort((a, b) => a.localeCompare(b));
    }
  });

  function dfs(comb, key) {
    let routes = map[key];

    if (comb.length === tickets.length) {
      return [...comb];
    }

    if (!routes || !routes.length) {
      return;
    }

    let result = [];

    for (let i = 0; i < routes.length; i++) {
      let route = routes[i];

      routes.splice(i, 1);
      comb.push(route);

      result = dfs(comb, route);

      routes.splice(i, 0, route);
      comb.pop();

      if (result) {
        return result;
      }
    }
  }

  let ans = dfs([], 'JFK', 0);

  ans.unshift('JFK');

  return ans;
};
